docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+        -- Perform formatting on T
+    end handler
+
+In this case, the compiler would identify that `M.FormatAsString<T>()` has a
+name that matches a generic handler declaration, and that it conforms (all
+generic parameters are specialised and all non-generic parameters are of the
+correct types).  Subsequent calls to `FormatAsString()` from *anywhere in the
+program* would then be able to access `M.FormatAsString<T>()` if the first
+argument of `FormatAsString()` conforms to `T`.
+
+The main drawback here is that Builder is optionally-typed, but it would be
+good to ensure that it's possible to write code where all types are specified
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end trait
+
+A trait definition is a block containing a list zero or more `method`
+definitions or `method type` declarations.  `method type` declarations take the
+same form as existing `handler type` declarations.  `method` definitions take
+the same form as existing `handler` declarations.
+
+    public trait type CanFormatAsString
+        method FormatAsString(in me) as string
+            return "<unknown>"
+        end method
+    end trait
+
+`method` definitions allow a trait's definition to include default
+implementations of the trait's methods.  Default implementations can call other
+methods of the trait or perform any operations valid for the generic `Self`
+type.
+
+Two additional symbols are defined within a trait definition:
+
+1. `MyType` is always equivalent to the type of the value that the method is
+   being called for.
+2. `me` must always be the first parameter a `method` definition or
+   `method type` declaration; it is equivalent to writing `me as MyType`.  The
+   `me` parameter must always be `in` or `inout`.
+
+### Implementing traits
+
+    public implementation of CanFormatAsString for T
+        method FormatAsString(in me) as string
+            variable tFormatted as string
+            -- Format me parameter
+            return tFormatted
+        end method
+    end implementation
+
+The `me` and `MyType` symbols are also defined inside implementation blocks.
+
+In order for an `implementation of U for T` to be valid, the following criteria
+must be fulfilled:
+
+1. Either `U` or `T` must be local, i.e. defined in the current module.  This
```

Comment:
Well developing a concrete specification of how versioning needs to work depends on knowing what it needs to account for ;)

A module import is a triple (module, version, symbol) - so a compiled module has a list of these.

Perhaps I should clarify my example above as I'm not sure your proposed resolution works in this case.

  - Miss Foo creates a type org.example.foo.Type@1
  - Mr Bar creates a trait org.example.bar.Trait@1 and implements it for all popular types including org.example.foo.Type@1
  - Mr Bar's trait becomes widely popular so lots of people use it
  - Miss Foo sees this and realizes that she can implement the trait much more efficiently in the module implementing Type@1 but obviously she cannot do so because either a trait has to implemented next to the type or next to the trait (you can't do both because it would result in a cyclic dependency).

So I'm wondering if a suitably designed way of compiling versioned modules solves this problem: Miss Foo creates org.example.foo.Type@2 and implements org.example.bar.Trait@1.

Thus:
  - org.example.foo@1 does not depend on org.example.bar@1
  - org.example.bar@1 depends on org.example.foo.Type@1
  - org.example.foo@2 depends on org.example.bar.Trait@1
  - org.example.bar@1 does *not* depend on org.example.foo@2 as it was compiled before the latter was created

Now, when Mr Bar comes to update his module he will get a cyclic dependency error - this will force him to version his module. All he has to do is move org.example.bar.Trait into a 'version 1' part which uses org.example.foo@1 - thus removing it from version 2.